Add universal equality

julienrf opened the pull request.

Ichoran commented on src/main/scala/strawman/collection/Iterable.scala:
```diff
@@ -132,6 +134,42 @@ trait IterableOps[+A] extends Any {
+  def sameElements[B >: A](that: Iterable[B]): Boolean = {
+  val these = this.iterator()
+  val those = that.iterator()
+  while (these.hasNext && those.hasNext)
```
Something is wonky about the indentation of this while loop. Also, any reason this shouldn't be a one-liner?  The extra lines just give an opportunity to get confused about indentation.

Ichoran commented on src/main/scala/strawman/collection/Iterable.scala:
```diff
@@ -132,6 +134,42 @@ trait IterableOps[+A] extends Any {
+  while (these.hasNext && those.hasNext)
+    if (these.next() != those.next())
+      return false
+    !these.hasNext && !those.hasNext
```
Logically this only needs to be `these.hasNext == those.hasNext`

julienrf replied:
Thanks!

I also fixed the indentation.

Ichoran commented on src/main/scala/strawman/collection/Iterable.scala:
```diff
@@ -132,6 +134,42 @@ trait IterableOps[+A] extends Any {
+    xs foreach { x =>
+      h = MurmurHash3.mix(h, x.##)
+      n += 1
+    }
```
Unless you have benchmarking results indicating that `foreach` is at least as fast as `iterator`, I would manually iterate this one to avoid needing to wrap `n` and `h` to pass into the closure.

julienrf replied:
I would be interested in having concrete numbers before doing this kind of changes that trade performance for readability.

I will wait for our benchmark test-kit.

odersky commented:
On Thu, Jan 5, 2017 at 8:08 PM, Julien Richard-Foy <notifications@github.com
> wrote:

> *@julienrf* commented on this pull request.
> ------------------------------
>
> In src/main/scala/strawman/collection/Iterable.scala
> <https://github.com/scala/collection-strawman/pull/12>:
>
> > +    IterableUtils.orderedHash(coll, IterableUtils.iterableSeed)
> +
> +}
> +
> +// Temporary: TODO move to MurmurHash3.scala
> +object IterableUtils {
> +
> +  val iterableSeed: Int = "Iterable".##
> +
> +  final def orderedHash(xs: Iterable[_], seed: Int): Int = {
> +    var n = 0
> +    var h = seed
> +    xs foreach { x =>
> +      h = MurmurHash3.mix(h, x.##)
> +      n += 1
> +    }
>
> I would be interested in having concrete numbers before doing this kind of
> changes that trade performance for readability.
>
I think it's better to be conservative here. The sad fact is that authors
of compilers and base libraries never get to enjoy the fruits of their
labor. They always have to be as close to the metal as possible, for the
benefit of everyone else. The problem with assuming the best and
banchmarking is that you don't know what the benchmarks should be. What are
typical loads? Is there even such a thing? Also, you need to make sure that
it works on all platforms including Android without a performance hit. So
unless you can check that it's actually the same bytecode being generated
it's better to assume the worst and write more.


I will wait for our benchmark test-kit.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scala/collection-strawman/pull/12>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAwlVu7rXR4UB5iwEip2li6my4t3u0Loks5rPOtLgaJpZM4Lad54>
> .
>



-- 

Martin Odersky
EPFL and Lightbend

julienrf commented:
Alright, I inlined the `foreach` call :)

odersky commented:
If we follow current collections, universal equality should be defined on `Seq` (and `Set` and `Map`), not on `Iterable`. I think that makes sense. Otherwise we get into awkwardness when we have to override Iterable's equality in Set and consequently have to deal with the binary method problem. Also, that way we can apply multiversal equality nicely, so that it will make no sense to compare a sequence with a set.

odersky commented:
So I would propose to move everything to the level of `collection.Seq`. Otherwise things look good to me.

julienrf commented:
Yes, I definitely agree.

julienrf commented:
Also, should we still use `canEqual`? I think we should.

sjrd commented:
I've never quite understood the real purpose of `canEqual`. I get it in theory, but I've never seen an actual use case.

szeiger commented:
I think we should err on the safe side with `Equals` and `canEqual` but I wouldn't expect it to make much of a difference in practice for collections.

julienrf commented:
I added a conservative `canEqual` implementation, like we have in 2.12.

Ichoran commented:
This looks fine but without a `Set` implementation we can't properly test it.

szeiger commented on src/main/scala/strawman/collection/Iterable.scala:
```diff
@@ -132,8 +132,29 @@ trait IterableOps[+A] extends Any {
 
   override def toString = s"$className(${mkString(", ")})"
+
+  def sameElements[B >: A](that: Iterable[B]): Boolean = {
```
Should this be in `IterableOps`? It doesn't really make sense for Sets and Maps with an undefined traversal order.

smarter replied:
See also: https://groups.google.com/forum/#!msg/scala-debate/je4vZbR3WbA/_fa5aNyTyuoJ, the method name is just wrong, I hope now is the chance to fix it.

julienrf replied:
You’re right, that makes no sens to inherit from that in `Set`. Is it ok to remove it from here? (it will not be backward compatible)

szeiger replied:
We could add a separate trait for collections with stable traversal order that can be extended by both `Seq` and `Sorted`.

szeiger commented on src/main/scala/strawman/collection/Iterable.scala:
```diff
@@ -132,8 +132,29 @@ trait IterableOps[+A] extends Any {
+    *  @return  `true`, if this $coll can possibly equal `that`, `false` otherwise. The test
+    *           takes into consideration only the run-time types of objects but ignores their elements.
+    */
+  def canEqual(that: Any): Boolean = true
```
I would also leave `canEqual` (and inheriting from `Equals`) out of `IterableOps` and implement it in the same places where `equals` is overridden.

julienrf replied:
That sounds reasonable too.

julienrf commented:
So. I moved the `equals` override to a new trait, `StableIterable`, which expects the traversal order to be stable. This trait is mixed into `collection.Seq` (and will be mixed into `collection.Sorted` too, when we will introduce this collection).

The difference with 2.12 is that `SortedSet(1, 2, 3) == List(1, 2, 3)` will return `true`. I’m not sure it is useful but since we support `ArrayBuffer(1, 2, 3) == List(1, 2, 3)` I see no reason to extend this to `Sorted` collections.

julienrf commented:
I rebased to master and added universal equality to `Set`.

I think we should wait for other collections to be actually implemented before going further…

szeiger commented:
Universal equality for `StableIterable` breaks transitivity. You have `HashSet(a,b) == TreeSet(a,b)` and `TreeSet(a,b) == Seq(a,b)` but not `HashSet(a,b) == Seq(a,b)`. Apart from that problem it looks good.

julienrf commented:
@szeiger Unless I’m wrong, if we mix `StableIterable` into `SortedSet` we will not have `HashSet(a, b) == TreeSet(a, b)` anymore.

szeiger commented:
True, but I doubt that is what we want. `HashSet(a, b) == TreeSet(a, b)` seems much more natural to me than `Seq(a, b) == TreeSet(a, b)`.

Ichoran commented:
Agreed.  That TreeSet happens to be in order is mostly implementation detail; it has little bearing on its core duty as a set, which is to be defined by its elements.  We also have the question about whether a TreeSet with a different ordering is a different set.  So is TreeSet(2,1) equal to TreeSet(1,2)?  If no, then TreeSet is more about ordering than about being a set.  If yes, then it's more about being a set.

(In those two cases, I mean ordering by `>` and `<` respectively.  If you don't supply an ordering, of course both should be the same regardless of how you build them.)

szeiger commented:
Comparing TreeSets with different ordering looks like another instance of the same problem. If you wanted to take ordering into account for equality of TreeSets, they couldn't be equal to other Sets anymore, otherwise you'd end up with the same transitivity problem.

Ichoran commented:
Right.  So going by LSP, if a TreeSet isa Set, then TreeSet must use set equality.

That said, I see nothing wrong with allowing `TreeSet` to implement `sameElementsInSameOrder` or whatever we call it and allowing that to check against Seqs.  But vanilla equals seems out.

odersky commented:
I agree that all sets should use set equality, order should not matter. That means we can merge `StableIterable` and `Seq` achieving some simplification, right?

szeiger commented:
`StableIterable` is still useful because it provides `sameElements` for both, Seqs and ordered Sets. The new implementation doesn't allow you to compare them with each other anymore. We could move `sameElements` up to `Iterable` but then we end up with the same unclear semantics as in 2.12. I would prefer to make the contract explicit with a `StableIterable` trait rather than renaming it to `sameElementsInSameOrder`.

odersky commented:
> I would prefer to make the contract explicit with a StableIterable trait rather than renaming it to sameElementsInSameOrder.

Are there other methods that should go into `StableIterable`?

sjrd commented:
Is there even a known use case for `sameElementsInSameOrder` between `Seq`s and `SortedSet`s (whatever the name of that method)? Shouldn't we leave it out entirely unless and until we have concrete evidence that it is useful in practice?

Ichoran commented:
There is also the possibility of simply writing `mySortedSet.iterator sameElements mySeq.iterator`.

szeiger commented:
Yeah, it's probably too much of a corner case to consider. As long as `Iterator` has a way to compare elements we should be good.

julienrf commented:
I moved `sameElements` to `Iterator` (and kept the definition in `Seq` too, which now delegates to the one of `Iterator`)
